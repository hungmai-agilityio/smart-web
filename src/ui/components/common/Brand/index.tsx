import Link from 'next/link';
import Image from 'next/image';

interface BrandProps {
  logo: string;
  url?: string;
}

export const Brand = ({ logo, url = '/' }: BrandProps) => {
  return (
    <Link href={url} className="cursor-pointer">
      <Image src={logo} alt="logo" width={160} height={84} priority />
    </Link>
  );
};
