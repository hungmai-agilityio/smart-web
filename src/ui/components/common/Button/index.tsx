'use client';

import { ButtonHTMLAttributes } from 'react';
import { clsx } from 'clsx';

// Constants
import { TYPE } from '@/constants';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: TYPE.PRIMARY | TYPE.SECOND;
  styles?: string;
}

export const Button = ({ name, variant, styles, disabled }: ButtonProps) => {
  const typeClasses = {
    [TYPE.PRIMARY]: 'bg-second',
    [TYPE.SECOND]:
      'bg-white border-primary border-2 border-primary hover:bg-second hover:border-0'
  };

  const typeClass = typeClasses[variant || TYPE.PRIMARY];

  const handleClick = () => {
    alert('We will update feature later');
  };

  return (
    <button
      className={clsx(
        'text-primary px-2 rounded-full font-light truncate',
        'w-24 h-6 text-xs sm:w-28 sm:h-8 sm:text-sm md:w-36 md:h-10 md:text-base lg:w-52 lg:h-12 lg:text-lg',
       disabled && 'cursor-not-allowed hover:bg-gray-400',
        typeClass,
        styles
      )}
      onClick={handleClick}
      disabled={disabled}
    >
      {name}
    </button>
  );
};
