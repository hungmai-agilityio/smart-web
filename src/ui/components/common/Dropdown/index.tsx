'use client';
import { useRef, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretDown } from '@fortawesome/free-solid-svg-icons';
import { useOutsideClick } from '@/hooks/useOutsideClick';

export interface ItemProps {
  label: string;
  value: string;
}

interface DropdownProps {
  items: ItemProps[];
  onSelect?: (value: string) => void;
}

export const Dropdown = ({ items, onSelect }: DropdownProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedItem, setSelectedItem] = useState<ItemProps>(items[0]);

  const dropdownRef = useRef<HTMLDivElement>(null);

  const toggleDropdown = () => {
    setIsOpen((prev) => !prev);
  };

  const handleItemClick = (item: ItemProps) => {
    setSelectedItem(item);
    setIsOpen(false);
    if (onSelect) onSelect(item.value);
  };

  useOutsideClick({
    ref: dropdownRef,
    handler: () => setIsOpen(false)
  });

  return (
    <div ref={dropdownRef} className="relative inline-block text-left w-36">
      <button
        type="button"
        className="flex items-center justify-between w-full px-3 py-2 border border-primary rounded-full md:text-base text-x-xs uppercase"
        onClick={toggleDropdown}
      >
        {selectedItem.label}
        <FontAwesomeIcon
          icon={faCaretDown}
          className={`ml-2 transition-transform ${isOpen ? 'rotate-180' : 'rotate-0'}`}
        />
      </button>

      <div
        className={`absolute right-0 w-full mt-1 bg-white border border-gray-300 rounded-md shadow-lg z-10 transition-transform duration-300 ease-in-out ${
          isOpen ? 'opacity-100 translate-y-0' : 'opacity-0 translate-y-2'
        }`}
      >
        {items.map((item) => (
          <button
            key={item.value}
            className="w-full text-left px-4 py-2 hover:bg-gray-100 md:text-base text-xs uppercase"
            onClick={handleItemClick.bind(null, item)}
          >
            {item.label}
          </button>
        ))}
      </div>
    </div>
  );
};
