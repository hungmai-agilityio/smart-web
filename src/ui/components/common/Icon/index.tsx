import { clsx } from 'clsx';

// Font Awesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface IconProp {
  icon: IconDefinition;
  quantity?: number;
  styles?: string;
}
export const Icon = ({ icon, quantity, styles }: IconProp) => {
  return (
    <div className={clsx('flex gap-3 items-center', styles)}>
      <FontAwesomeIcon icon={icon} />

      <p>{quantity}</p>
    </div>
  );
};
