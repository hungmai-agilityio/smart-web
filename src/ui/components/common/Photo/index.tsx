'use client';
import { ReactNode, useState } from 'react';
import { clsx } from 'clsx';
import Image, { ImageProps } from 'next/image';

interface PhotoProps extends ImageProps {
  children?: ReactNode;
  styles?: string;
  isCircle?: boolean;
  fixedSize?: { width: number; height: number };
  autoSize?: boolean;
}

export const Photo = ({
  src,
  alt,
  children,
  styles,
  isCircle,
  fixedSize,
  autoSize,
  ...res
}: PhotoProps) => {
  const [hovered, setHovered] = useState(false);

  const toggleHover = () => setHovered(!hovered);

  const containerStyles = clsx('relative', styles, {
    'rounded-full overflow-hidden border-4 border-gray-300 hover:border-second':
      isCircle,
    'w-full h-full': autoSize,
    [`w-[${fixedSize?.width}px] h-[${fixedSize?.height}px]`]:
      fixedSize && !autoSize
  });

  return (
    <div
      className={containerStyles}
      onMouseEnter={toggleHover}
      onMouseLeave={toggleHover}
    >
      <Image
        src={src}
        alt={alt}
        fill={autoSize}
        width={fixedSize?.width}
        height={fixedSize?.height}
        style={{ objectFit: 'cover' }}
        className={clsx('transition-opacity duration-300', {
          'opacity-50': hovered && children
        })}
        {...res}
      />
      {hovered && children && (
        <div className="absolute inset-0 flex items-center bg-black bg-opacity-50 text-white">
          {children}
        </div>
      )}
    </div>
  );
};
