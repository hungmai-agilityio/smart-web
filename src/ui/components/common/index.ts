export * from '@/ui/components/common/Button';
export * from '@/ui/components/common/Dropdown';
export * from '@/ui/components/common/Icon';
export * from '@/ui/components/common/Photo';
export * from '@/ui/components/common/Text';
export * from '@/ui/components/common/StatsRow';
