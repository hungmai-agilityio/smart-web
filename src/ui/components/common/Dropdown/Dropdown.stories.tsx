import { Meta, StoryObj } from '@storybook/react';
import { itemDropdown } from '@/mocks';

// Components
import { Dropdown } from '@/ui/components/common/Dropdown';

const meta: Meta<typeof Dropdown> = {
  component: Dropdown,
  tags: ['autodocs'],
  argTypes: {
    onSelect: {
      description:
        'Callback function that is called when an item is selected. Receives the value of the selected item.'
    },
    items: {
      description:
        'Array of items where each item contains a label to be displayed and a value to be used in logic.'
    }
  }
};

export default meta;

export const Default: StoryObj<typeof Dropdown> = {
  args: {
    items: itemDropdown
  }
};
