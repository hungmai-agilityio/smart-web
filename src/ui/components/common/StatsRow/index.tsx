import { clsx } from 'clsx';

//Components
import { Photo } from '@/ui/components/common';

interface StatsRowProps {
  views: number;
  favorite: number;
  comment: number;
  isColumn?: boolean;
  viewIcon: string;
  favoriteIcon: string;
  commentIcon: string;
}

export const StatsRow = ({
  views,
  favorite,
  comment,
  isColumn,
  viewIcon,
  favoriteIcon,
  commentIcon
}: StatsRowProps) => {
  return (
    <div
      className={clsx('flex px-3 justify-between md:items-center w-full', {
        'flex-col md:hidden gap-2': isColumn
      })}
    >
      <div className="flex gap-1 items-center">
        <div
          className={clsx('w-icon-xs h-icon-xxs', {
            'w-7 h-icon-sm': isColumn
          })}
        >
          <Photo src={viewIcon} alt="view icon" autoSize />
        </div>
        <p
          className={clsx(
            'xl:w-10 w-8 truncate overflow-hidden',
            isColumn && 'w-full text-white'
          )}
        >
          {views}
        </p>
      </div>

      <div className="flex gap-1 items-center">
        <div
          className={clsx('w-icon-xxs h-icon-xxs', {
            'w-5 h-icon-xs': isColumn
          })}
        >
          <Photo src={favoriteIcon} alt="likes icon" autoSize />
        </div>
        <p
          className={clsx(
            'xl:w-10 w-8 truncate overflow-hidden',
            isColumn && 'w-full text-white'
          )}
        >
          {favorite}
        </p>
      </div>

      <div className="flex gap-1 items-center">
        <div
          className={clsx('w-icon-xxs h-icon-xxs', {
            'w-5 h-icon-xs': isColumn
          })}
        >
          <Photo src={commentIcon} alt="comments icon" autoSize />
        </div>
        <p
          className={clsx(
            'xl:w-10 w-8 truncate overflow-hidden',
            isColumn && 'w-full text-white'
          )}
        >
          {comment}
        </p>
      </div>
    </div>
  );
};
