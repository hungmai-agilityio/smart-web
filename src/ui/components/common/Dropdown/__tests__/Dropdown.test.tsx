import { render, fireEvent, screen } from '@testing-library/react';

// Mocks
import { itemDropdown } from '@/mocks';

// Components
import { Dropdown } from '@/ui/components/common/Dropdown';

describe('Dropdown Component', () => {
  const onSelect = jest.fn();

  beforeEach(() => {
    jest.fn().mockClear();
  });

  it('renders correctly with default item', () => {
    render(<Dropdown items={itemDropdown} />);
    const [selectedItem] = screen.getAllByText('Option 1');
    expect(selectedItem).toBeInTheDocument();
  });

  it('opens the dropdown and selects an item', () => {
    render(<Dropdown items={itemDropdown} onSelect={onSelect} />);

    const [dropdownButton] = screen.getAllByText('Option 1');
    fireEvent.click(dropdownButton);

    const item2 = screen.getByText('Option 2');
    fireEvent.click(item2);

    const [selectedItem] = screen.getAllByText('Option 2');
    expect(selectedItem).toBeInTheDocument();
    expect(onSelect).toHaveBeenCalledWith('option2');
  });

  it('should be render match to snapshot', () => {
    const { container } = render(<Dropdown items={itemDropdown} />);
    expect(container).toMatchSnapshot();
  });
});
