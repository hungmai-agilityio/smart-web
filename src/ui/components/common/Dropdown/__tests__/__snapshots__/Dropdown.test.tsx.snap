// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component should be render match to snapshot 1`] = `
<div>
  <div
    class="relative inline-block text-left w-36"
  >
    <button
      class="flex items-center justify-between w-full px-3 py-2 border border-primary rounded-full md:text-base text-x-xs uppercase"
      type="button"
    >
      Option 1
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down ml-2 transition-transform rotate-0"
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="absolute right-0 w-full mt-1 bg-white border border-gray-300 rounded-md shadow-lg z-10 transition-transform duration-300 ease-in-out opacity-0 translate-y-2"
    >
      <button
        class="w-full text-left px-4 py-2 hover:bg-gray-100 md:text-base text-xs uppercase"
      >
        Option 1
      </button>
      <button
        class="w-full text-left px-4 py-2 hover:bg-gray-100 md:text-base text-xs uppercase"
      >
        Option 2
      </button>
      <button
        class="w-full text-left px-4 py-2 hover:bg-gray-100 md:text-base text-xs uppercase"
      >
        Option 3
      </button>
    </div>
  </div>
</div>
`;
